xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,290 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+    }
+    ErrorSpec error_spec{1e-4, 1e-4};
+    if (!root_sharded_dims.empty()) {
+      std::vector<int64_t> lower(root_dims_per_shard.size(), 0);
+      std::vector<int64_t> upper(root_dims_per_shard.begin(),
+                                 root_dims_per_shard.end());
+      for (const Literal& result : results) {
+        Literal ref_results_slice = ref_results[0].Slice(lower, upper);
+        EXPECT_TRUE(
+            LiteralTestUtil::Near(ref_results_slice, result, error_spec));
+        for (int m = root_sharded_dims.size() - 1; m >= 0; --m) {
+          if (upper[root_sharded_dims[m]] < root_dims[root_sharded_dims[m]]) {
+            upper[root_sharded_dims[m]] +=
+                root_dims_per_shard[root_sharded_dims[m]];
+            break;
+          } else {
+            upper[root_sharded_dims[m]] =
+                root_dims_per_shard[root_sharded_dims[m]];
+          }
+        }
+        std::transform(upper.begin(), upper.end(), root_dims_per_shard.begin(),
+                       lower.begin(), std::minus<int64_t>());
+      }
+    } else {
+      EXPECT_TRUE(
+          LiteralTestUtil::Near(ref_results[0], results[0], error_spec));
+    }
+  }
+
+ private:
+  // Execute the replicated case.
+  absl::StatusOr<std::vector<Literal>> ExecuteReplicated(
+      const std::string& hlo_text) {
+    // Create the unsharded reference case by removing the sharding metadata
+    // from the HLO string.
+    std::string hlo_text_ref = hlo_text;
+    RE2::GlobalReplace(&hlo_text_ref, R"(, sharding=\{devices=\[[0-9,]*\].*\})",
+                       "");
+    RE2::GlobalReplace(&hlo_text_ref, R"(, sharding=\{replicated\})", "");
+
+    HloModuleConfig ref_config = GetModuleConfigForTest(/*replica_count=*/1);
+    auto ref_opts = GetDebugOptionsForTest();
+    ref_opts.set_xla_gpu_enable_triton_gemm(false);
+    ref_config.set_debug_options(ref_opts);
+    ref_config.set_num_partitions(1);
+    TF_ASSIGN_OR_RETURN(auto ref_module,
+                        ParseAndReturnVerifiedModule(hlo_text_ref, ref_config));
+    const int64_t num_params =
+        ref_module->entry_computation()->num_parameters();
+
+    auto fake_args = xla::MakeFakeArguments(ref_module.get()).value();
+    std::vector<Literal*> ref_fake_ptrs(num_params);
+    for (int i = 0; i < num_params; ++i) {
+      ref_fake_ptrs[i] = &fake_args[i];
+    }
+
+    DeviceAssignment ref_assn(/*replica_count=*/1,
+                              /*computation_count=*/1);
+    ref_assn(0, 0) = 0;
+    return HloTestBase::ExecuteReplicated(std::move(ref_module), ref_fake_ptrs,
+                                          /*num_replicas=*/1, &ref_assn,
+                                          /*run_hlo_passes=*/true,
+                                          /*use-threads=*/true);
+  }
+
+  // Execute the sharded case.
+  absl::StatusOr<std::vector<Literal>> ExecuteSharded(
+      const std::string& hlo_text, int64_t num_partitions,
+      int64_t num_replicas) {
```

Comment:
I think num_replicas is always 1 in this case. You are only dealing with multiple devices in the partition Dimension. You can just pass in a constant  1, e.g. `/*replica_count=*/1`